src/parser/mod.rs

```diff
@@ -1427,6 +1426,112 @@ impl<'a> Parser<'a> {
         }
     }
 
+    /// Try to parse an [Expr::CompoundExpr] like `a.b.c` or `a.b[1].c`.
+    /// If all the fields are `Expr::Identifier`s, return an [Expr::CompoundIdentifier] instead.
+    /// If only the root exists, return the root.
+    /// If self supports [Dialect::supports_partiql], it will fall back when occurs [Token::LBracket] for JsonAccess parsing.
+    pub fn parse_compound_expr(
+        &mut self,
+        root: Expr,
+        mut chain: Vec<AccessField>,
+    ) -> Result<Expr, ParserError> {
+        let mut ending_wildcard: Option<TokenWithSpan> = None;
+        let mut ending_lbracket = false;
+        while self.consume_token(&Token::Period) {
+            let next_token = self.next_token();
+            match next_token.token {
+                Token::Word(w) => {
+                    let expr = Expr::Identifier(w.to_ident(next_token.span));
+                    chain.push(AccessField::Expr(expr));
+                    if self.consume_token(&Token::LBracket) {
+                        if self.dialect.supports_partiql() {
+                            ending_lbracket = true;
+                            break;
+                        } else {
+                            self.parse_multi_dim_subscript(&mut chain)?
+                        }
+                    }
+                }
+                Token::Mul => {
+                    // Postgres explicitly allows funcnm(tablenm.*) and the
+                    // function array_agg traverses this control flow
+                    if dialect_of!(self is PostgreSqlDialect) {
+                        ending_wildcard = Some(next_token);
+                        break;
+                    } else {
+                        return self.expected("an identifier after '.'", next_token);
+                    }
+                }
+                Token::SingleQuotedString(s) => {
+                    let expr = Expr::Identifier(Ident::with_quote('\'', s));
+                    chain.push(AccessField::Expr(expr));
+                }
+                _ => {
+                    return self.expected("an identifier or a '*' after '.'", next_token);
+                }
+            }
+        }
+
+        // if dialect supports partiql, we need to go back one Token::LBracket for the JsonAccess parsing
+        if self.dialect.supports_partiql() && ending_lbracket {
+            self.prev_token();
+        }
+
+        if let Some(wildcard_token) = ending_wildcard {
```

Comment:
I'm wondering if it is required that `parse_compound_expr` track the wildcard and other cases that follow. I imagined the behavior somewhat similar to what the doc comment suggests, in that the method only looks to parse a chain if one exists (it looks like the function and outer join parsing is duplicated here which doesnt seem ideal).

Would it be feasible to have this function only take in the root and then look to parse only the chain? e.g. roughly
```rust
let mut chain = vec![];
loop {
  match self.peek().token {
    Token::Period => {
      self.next_token(); // `.`
      if let Some(expr) = self.maybe_parse(Self::parse_sub_expr(
          // Guessing a bit offhand that involving precedence here
          // could help avoid the need to flatten afterwards. I _think_ that was
          // the intent with `exprs_to_idents`? idea is that in this context,
          // the dot is essentially an operator similar to `[]`
          self.dialect.prec_value(Percedence::Dot)
      ))? {
        chain.push(AccessField::Expr(expr));
      } else {
        self.prev_token(); // `.`
        break
      }
    }
    Token::LBracket => {
      // We can probably let the subscript method handle the `[` and `]`
      if let Some(subscript) = self.maybe_parse(Self::parse_subscript)? {
        chain.push(AccessField::Subscript(subscript));
      } else {
        break
      }
    }
  }
}
```

Reply:
I guess it's hard to do it 🤔  because the structure of `QualifiedWildcard` is
```
    QualifiedWildcard(ObjectName, AttachedToken),
```
It's similar to `CompoundIdentifier` and `CompoundExpr`. All of them will consume all the dots.

Reply:
Ah right indeed `CompoundExpr` is a superset of `CompoundIdentifier` so there likely wouldn't be a way to avoid that `exprs_to_idents` flattening, like in the scenario where a `foo.bar.baz` expr shows up. That case seems reasonable to flatten since we'd want to keep the representations distinct!

The QualifiedWildcard I think could be fine this same loop as above can handle it I imagine if we peek the next token after the period? i.e if we consume a `.*` then we build up the wildcard expr as we do today with the root+chain consumed so far. So that essentially the loop parses the chain delimited by period and the `CompoundIdentifier` vs `QualifiedWildcard` are the same scenarios only differing in the terminating token

Reply:
@iffyio 
I feel the idea of this loop is good but I found a problem 🤔 . If we try to invoke `parse_sub_expr` recursively, we will still get the nested representation. Consider the following cases:

## The representation of abc[1].f1.f2
The last field is a `CompoundIdentifier`.
```
&root = Identifier(
    Ident {
        value: "abc",
        quote_style: None,
        span: Span(Location(0,0)..Location(0,0)),
    },
)
&chain = [
    Subscript(
        Index {
            index: Value(
                Number(
                    "1",
                    false,
                ),
            ),
        },
    ),
    Dot(
        CompoundIdentifier(
            [
                Ident {
                    value: "f1",
                    quote_style: None,
                    span: Span(Location(0,0)..Location(0,0)),
                },
                Ident {
                    value: "f2",
                    quote_style: None,
                    span: Span(Location(0,0)..Location(0,0)),
                },
            ],
        ),
    ),
]
```
## The representation of abc[1].f1.f2[1]
The last field is a `CompoundFieldAccess` with another chain.
```
&root = Identifier(
    Ident {
        value: "abc",
        quote_style: None,
        span: Span(Location(0,0)..Location(0,0)),
    },
)
&chain = [
    Subscript(
        Index {
            index: Value(
                Number(
                    "1",
                    false,
                ),
            ),
        },
    ),
    Dot(
        CompoundFieldAccess {
            root: Identifier(
                Ident {
                    value: "f1",
                    quote_style: None,
                    span: Span(Location(0,0)..Location(0,0)),
                },
            ),
            access_chain: [
                Dot(
                    CompoundFieldAccess {
                        root: Identifier(
                            Ident {
                                value: "f2",
                                quote_style: None,
                                span: Span(Location(0,0)..Location(0,0)),
                            },
                        ),
                        access_chain: [
                            Subscript(
                                Index {
                                    index: Value(
                                        Number(
                                            "1",
                                            false,
                                        ),
                                    ),
                                },
                            ),
                        ],
                    },
                ),
            ],
        },
    ),
]
```
It seems we should consider `CompoundIdentifier` and `CompoundFieldAccess` as a kind of the nested struct. 
Ideally, the access_cahin shouldn't contain another nested struct. 
The `parse_subexpr` shouldn't produce the nested struct 🤔. It makes the thing more complicated.
I don't prefer to do It in this PR. The side effects will be too large. Maybe we can have a follow-up PR to refactor here.
WDYT?